File: Legacy/Leetcode/53MaximumSubarray.py
# Find the contiguous subarray within an array (containing at least one number) which has the largest sum.
#
# For example, given the array [-2,1,-3,4,-1,2,1,-5,4],
# the contiguous subarray [4,-1,2,1] has the largest sum = 6.

import sys

def merge(nums, mid):
    max_left = - sys.maxint
    max_right = - sys.maxint
    sum_left = 0
    sum_right = 0
    for i in range(mid + 1, len(nums)):
        sum_right += nums[i]
        if sum_right > max_right:
            max_right = sum_right
    for i in range(mid, -1, -1):
        sum_left += nums[i]
        if sum_left > max_left:
            max_left = sum_left
    return max_left + max_right

def max_sub_array(nums):
    if len(nums) == 1:
        return nums[0]
    else:
        start = 0
        end = len(nums) - 1
        mid = start + (end - start) / 2
        return max(max_sub_array(nums[start : mid + 1]), max_sub_array(nums[mid + 1 : end + 1]), merge(nums, mid))

print max_sub_array([-2,1,-3,4,-1,2,1,-5,4])

File: C++/Array/53. Maximum Subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        
        int maxVal = INT_MIN;
        for (int n : nums) maxVal = max(maxVal, n);
        if (maxVal <= 0) return maxVal;
        
        int res = 0;
        int posBlock = 0, negBlock = 0, merge = 0;
        int i = 0;
        while (i < nums.size()) {
            if (nums[i] > 0) {
                while (i < nums.size() && nums[i] > 0) {
                    posBlock += nums[i++];
                }
                res = max(res, posBlock);
            }
            else {
                negBlock = 0;
                while (i < nums.size() && nums[i] <= 0) {
                    negBlock += nums[i++];
                }
                merge = posBlock + negBlock;
                posBlock = merge > 0 ? merge : 0;
            }
        }
        return res;
    }
};